src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +45,337 @@ public static String newChallenge(
+    if (signersFound.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction not signed by any client signer.");
+    }
+
+    // Confirm all signatures were consumed by a signer.
+    if (signersFound.size() != transaction.getSignatures().size() - 1) {
+      throw new InvalidSep10ChallengeException("Transaction has unrecognized signatures.");
+    }
+
+    return signersFound;
+  }
+
+  /**
+   * Verifies that for a SEP-0010 challenge transaction
+   * all signatures on the transaction are accounted for and that the signatures
+   * meet a threshold on an account. A transaction is verified if it is signed by
+   * the server account, and all other signatures match a signer that has been
+   * provided as an argument, and those signatures meet a threshold on the
+   * account.
+   *
+   * @param challengeXdr    SEP-0010 transaction challenge transaction in base64.
+   * @param serverAccountId Account ID for server's account.
+   * @param network         The network to connect to for verifying and retrieving.
+   * @param threshold       The threshold on the client account.
+   * @param signers         The signers of client account.
+   * @return a list of signers that were found is returned, excluding the server account ID.
+   * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static Set<String> verifyChallengeTransactionThreshold(String challengeXdr, String serverAccountId, Network network, int threshold, Set<Signer> signers) throws InvalidSep10ChallengeException, IOException {
+    if (signers == null || signers.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No signers provided.");
+    }
+
+    Map<String, Signer> accountIdSignerMap = new HashMap<String, Signer>();
+    for (Signer signer : signers) {
+      accountIdSignerMap.put(signer.getKey(), signer);
+    }
+
+    Set<String> signersFound = verifyChallengeTransactionSigners(challengeXdr, serverAccountId, network, accountIdSignerMap.keySet());
+
+    int weight = 0;
+    for (String signer : signersFound) {
+      if (!accountIdSignerMap.containsKey(signer)) {
+        continue;
+      }
+      weight += accountIdSignerMap.get(signer).getWeight();
+    }
+
+    if (weight < threshold) {
+      throw new InvalidSep10ChallengeException(String.format("Signers with weight %d do not meet threshold %d.", weight, threshold));
+    }
+
+    return signersFound;
+  }
+
+  private static Set<String> verifyTransactionSignatures(Transaction transaction, Set<String> signers) throws InvalidSep10ChallengeException {
+    List<DecoratedSignature> signatures = transaction.getSignatures();
+    if (signatures.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction has no signatures.");
+    }
+
+    byte[] txHash = transaction.hash();
+
+    // find and verify signatures
```

Comment:
I think it's possible to simplify this using a Multimap:

```java
    // find and verify signatures
    Set<String> signersFound = new HashSet<String>();
    Multimap<SignatureHint, Signature> signatures = HashMultimap.create();
    for (DecoratedSignature decoratedSignature : transaction.getSignatures()) {
      signatures.put(decoratedSignature.getHint(), decoratedSignature.getSignature());
    }

    for (String signer : signers) {
      KeyPair keyPair = KeyPair.fromAccountId(signer);
      SignatureHint hint = keyPair.getSignatureHint();

      for (Signature signature : signatures.get(hint)) {
        if (keyPair.verify(txHash, signature.getSignature())) {
          signersFound.add(signer);
          // explicitly ensure that a transaction signature cannot be
          // mapped to more than one signer
          signatures.remove(hint, signature);
          break;
        }
      }
    }
```

Reply:
This looks great, but `Multimap` is not a built-in data type. Should we use a third-party data type when it is not necessary?

Reply:
@overcat great point! Multimap comes https://github.com/google/guava . We already have a dependency on guava ( https://github.com/stellar/java-stellar-sdk/blob/master/build.gradle#L45-L46 ) so this change would not require any new dependencies. Note that the `Optional` implementation we're using in the java sdk also comes from guava.

Reply:
OK, changed.